libraries/AP_Torqeedo/AP_Torqeedo.cpp

```diff
@@ -128,22 +160,70 @@ void AP_Torqeedo::thread_main()
 
         // parse incoming characters
         uint32_t nbytes = MIN(_uart->available(), 1024U);
-        bool motor_speed_request_received = false;
         while (nbytes-- > 0) {
             int16_t b = _uart->read();
             if (b >= 0 ) {
                 if (parse_byte((uint8_t)b)) {
                     // request received to send updated motor speed
-                    motor_speed_request_received = true;
+                    if (_type == ConnectionType::TYPE_TILLER) {
+                        _send_motor_speed = true;
+                    }
                 }
             }
         }
 
         // send motor speed
-        if (motor_speed_request_received) {
-            send_motor_speed_cmd();
+        if (safe_to_send()) {
+            // if connected to motor send motor speed every 0.5sec
+            if (_type == ConnectionType::TYPE_MOTOR &&
+                (AP_HAL::micros() - _last_send_motor_us > TORQEEDO_SEND_MOTOR_SPEED_INTERVAL_US)) {\
+                _send_motor_speed = true;
+            }
+
+            // send motor speed
+            if (_send_motor_speed) {
+                send_motor_speed_cmd();
+                _send_motor_speed = false;
+            }
         }
+
+        // logging and debug output
+        log_and_debug();
+    }
+}
+
+// returns true if communicating with the motor
+bool AP_Torqeedo::healthy()
+{
+    if (!_initialised) {
+        return false;
+    }
+    {
+        // healthy if both receive and send have occurred in the last 3 seconds
+        WITH_SEMAPHORE(_last_healthy_sem);
```

Comment:
I went back and forth on this and I think we do actually need the semaphore because otherwise a call from the front end (main loop thread) could to healthy() could end up executing the "now_ms = AP_HAL::millis()" just before the backend thread updates, "_last_received_ms" and we could end up with a huge negative number for "(now_ms - _last_received_ms)".

Reply:
As discussed, there is an alternative fix which is to change the ordering of the lines to avoid the semaphore but we decided for now that using the semaphore is acceptable and is the safer option because compilers could theoretically move the order of instructions.

Reply:
We decided to leave this as is.....